pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -10,27 +13,188 @@ import (
 	"k8s.io/kubernetes/pkg/serviceaccount"
 
 	scopeauthorizer "github.com/openshift/origin/pkg/authorization/authorizer/scope"
+	osclient "github.com/openshift/origin/pkg/client"
 	oauthapi "github.com/openshift/origin/pkg/oauth/api"
 	"github.com/openshift/origin/pkg/oauth/registry/oauthclient"
+	routeapi "github.com/openshift/origin/pkg/route/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
 )
 
 const (
-	OAuthRedirectURISecretAnnotationPrefix = "serviceaccounts.openshift.io/oauth-redirecturi."
-	OAuthWantChallengesAnnotationPrefix    = "serviceaccounts.openshift.io/oauth-want-challenges"
+	OAuthWantChallengesAnnotationPrefix = "serviceaccounts.openshift.io/oauth-want-challenges"
+
+	// Base prefix for redirect URI annotations
+	OAuthRedirectModelAnnotationPrefix = "serviceaccounts.openshift.io/oauth-redirect"
+
+	// Prefix used for statically specifying redirect URIs for a service account via annotations
+	// The value can be partially supplied with the dynamic prefix to override the resource's defaults
+	OAuthRedirectModelAnnotationURIPrefix = OAuthRedirectModelAnnotationPrefix + "uri."
+
+	// Prefix used for dynamically specifying redirect URIs using resources for a service account via annotations
+	OAuthRedirectModelAnnotationReferencePrefix = OAuthRedirectModelAnnotationPrefix + "ref."
+
+	RouteKind = "Route"
+	// TODO add ingress support
+	// IngressKind = "Ingress"
 )
 
+var modelPrefixes = []string{
+	OAuthRedirectModelAnnotationURIPrefix,
+	OAuthRedirectModelAnnotationReferencePrefix,
+}
+
+// namesToObjMapperFunc is linked to a given GroupKind.
+// Based on the namespace and names provided, it builds a map of resource name to redirect URIs.
+// The redirect URIs represent the default values as specified by the resource.
+// These values can be overridden by user specified data.
+type namesToObjMapperFunc func(namespace string, names sets.String) map[string]redirectURIList
+
+var emptyGroupKind = unversioned.GroupKind{} // Used with static redirect URIs
+var routeGroupKind = routeapi.SchemeGroupVersion.WithKind(RouteKind).GroupKind()
+
+// TODO add ingress support
+// var ingressGroupKind = routeapi.SchemeGroupVersion.WithKind(IngressKind).GroupKind()
+
 type saOAuthClientAdapter struct {
 	saClient     kclient.ServiceAccountsNamespacer
 	secretClient kclient.SecretsNamespacer
+	routeClient  osclient.RoutesNamespacer
+	// TODO add ingress support
+	//ingressClient ??
 
 	delegate    oauthclient.Getter
 	grantMethod oauthapi.GrantHandlerType
 }
 
+// redirectObjectReference holds the resource components needed to reference an object in the same namespace
+type redirectObjectReference struct {
```

Comment:
if we're encoding this to annotations, we should follow the pattern of SerializedReference/ObjectReference:

1. you'll need two API types - one for the data you care about (group/kind/name), and one that is registered as a top-level type (with TypeMeta, so we can encode/decode the data in the annotation using standard object encoders/decoders)
2. generate the annotation by encoding the top-level object (see how createdByRefJson is created upstream)
3. parse the annotation by decoding the annotation into an object (see how creatorRefRaw is decoded into an object upstream)